Assorted fixes

dagwieers opened the pull request:
This PR includes:
- Fix a bug caused by the dot in the return signal name
- Replace RECEIVER global with static variable
- Replace getAddonInfo() call with _addon_id() using static variable
- Fix SPDX to LGPL-2.1-or-later
- ~Switch to using unicode_literals for Python 2~
- Add _to_unicode() function
- Remove unused CallHandler.data
- Use selective imports
    
And also cosmetic/testing improvements:
- Add docstrings to all methods and functions
- Remove disabled import-error and missing-docstring pylint tests
- A codecov.yml file for code coverage
- Add coverage support to Travis config
- Small fixes to setup.py and tox config
- Allow PYTHON to be set for Makefile
- Unit test fix
- Use snake_case for internal functions (no impact)
- Add tests for registerSlot(), sendSignal() and unregisterSlot()
- Add tests for makeCall(), registerCall() and returnCall()

See Travis for test results:
https://travis-ci.org/dagwieers/script.module.addon.signals
And Codecov for coverage results (we have 99% testing coverage!):
https://codecov.io/gh/dagwieers/script.module.addon.signals/branch/assorted-fixes/commits

dagwieers committed:
Assorted fixes

This PR includes:
- Fix SPDX to LGPL-2.1-or-later
- Switch to using unicode_literals for Python 2
- Add _to_unicode() function
- Replace RECEIVER global with static variable
- Replace getAddonInfo() call with _addon_id() using static variable
- Remove unused CallHandler.data

And also cosmetic/testing improvements:
- Add docstrings to all methods and functions
- Remove disabled import-error and missing-docstring pylint tests
- A codecov.yml file for code coverage
- Add coverage support to Travis config
- Small fixes to setup.py and tox config
- Allow PYTHON to be set for Makefile
- Unit test fix
- Use snake_case for internal functions (no impact)

dagwieers commented:
~I know this is quite a lot of unrelated changes. The only impacting change in this list is the move to using unicode_literals for Python 2 and related simplification to the code. That will need to be tested in Kodi with existing add-ons.~ *I removed this for now*

Let me know what parts you like me to split up in different PRs.

dagwieers committed:
Avoid loading libraries at import time

This reduces the overhead of using AddonSignals at import time (as is
most impacting on slower devices like Raspberry Pi).

On Kodi all imports are performed at startup or at every invocation,
even when not used.

dagwieers commented:
@ruuk There are two additional things I would like you to review specifically:
- The docstrings for all methods
  - I am not that familiar with the terminology of AddonSignals
- Especially the `registerCall()`, `returnCall()` and `makeCall()` functions
  - not sure why we have them (backward compatiblity?) and how they ought to be used
  - I added integration tests, feel free to let me know what from those tests are useless

CastagnaIT commented:
a curiosity question, i have seen that in all the projects you keep wrong the commas on the descriptions of the methods, why do you insert them all wrong?
so you use the wrong ''' instead of """

dagwieers commented:
cc @mediaminister I always appreciate your input :wink:

CastagnaIT commented on lib/AddonSignals.py:
```diff
@@ -1,135 +1,162 @@
 # -*- coding: utf-8 -*-
+# GNU Lesser General Public License v2.1 (see COPYING or https://www.gnu.org/licenses/gpl-2.1.txt)
```
this not follow SPDX convention

dagwieers replied:
Comments in code do not use SPDX. SPDX identifiers are used for computer-processed fields not for human-processed information (license one-liners are intended for humans).

CastagnaIT replied:
I don't understand what you mean, but this does not follow the SPDX directives at all, if you do a test with various check software will not be approved.

You can see how to do in netflix repo,
even licenses should not be written with links, but placed in the folder LICENSES,
otherwise it is useless to follow the SPDX directives...

dagwieers replied:
Make a new PR for your change if you like. I am not interested to take it this far.

The license is in LICENSE.txt, I don't see the point of an additional LICENSES/ directory if we have a single license anyway. (The license is 5x larger than the library, so having the license in the source tree twice is unneeded and unwanted IMO).

Again, conventions are nice, up to a point and then they become pedantic :smile:

CastagnaIT replied:
it is addon checker, devs want to go in this way...
agree that this project is big like a flea

dagwieers replied:
AFAIK this was only related to addon.xml, not code comments in your source files.

dagwieers commented:
> so you use the wrong ''' instead of """

Why is using ''' wrong? There is no convention that places one before the other.

In Python I tend to use single quotes unless double quotes are required. I prefer single quotes because they are easier to read (less fluffy) then double quotes.

CastagnaIT commented:
so understand, i prefer to use the correct ones otherwise PEP8 and pycharm complains...

dagwieers commented:
> so understand, i prefer to use the correct ones otherwise PEP8 and pycharm complains...

Well, PEP8 has no convention on single or double quotes. And if pycharm complains to you for this, you may want to configure it correctly :wink: An editor should not force you to do things that don't matter.

Before you realize it, you start telling people they do something wrong when it is just their editor being picky on them :wink:

CastagnaIT commented:
i go to pull Pycharm's hairs

dagwieers commented:
@CastagnaIT In itself I don't mind more conventions (PEP8 is mostly great) but if personal preferences are enforced by editors, you may get in situations where Python editors have conflicting ideas of what is preferred. Whereas IMO conventions are best when they improve readability and avoid common issues.

That said I do prefer it if code is consistent, but I am flexible to what this consistency entails.

dagwieers commented:
And I just discovered that while PEP8 doesn't enforce single or double quotes per se, PEP257 does have a convention for triple double quotes for docstrings: https://www.python.org/dev/peps/pep-0257/

I guess that is why pyCharm is enforcing this :disappointed: So I guess it would be better to use triple double quotes eventually. (And this thread wasn't a waste of time after all :wink:)

CastagnaIT commented:
I seemed to have read somewhere but i didn't remember where...
I was just curious about your work, it wasn't to force you to do it

dagwieers committed:
Replace triple quotes by triple double quotes

dagwieers commented:
Damn, I really dislike triple double quotes with no spaces. Less readable IMO :roll_eyes:

dagwieers committed:
Add integration tests for calls

dagwieers commented:
I removed the use of unicode_literals for Python 2.

Personally I think this is what we should be doing, but I cannot guarantee that this won't break older add-ons that did not migrate to Python 3 compatibility.

dagwieers commented:
Since @CastagnaIT is only picking on the unimportant bits, it must mean the important changes look good :wink:

CastagnaIT commented:
I have yet to see all the changes... I'm looking in pieces

dagwieers committed:
Clean up pylint some more

CastagnaIT commented on lib/AddonSignals.py:
```diff
@@ -1,135 +1,161 @@
+    return text
+
+
+class SignalReceiver(Monitor, object):
```
why add object?

dagwieers replied:
Because we need new-style classes in Python 2 in order to use `__subclasses__`. For Python 3 it wouldn't be needed as it would be a new-style class anyway.

dagwieers replied:
This is only needed for integration testing though, unfortunately.

CastagnaIT commented:
in the end no more questions all just cosmetic changes

ruuk commented:
>   * I am not that familiar with the terminology of AddonSignals
> * Especially the `registerCall()`, `returnCall()` and `makeCall()` functions
>   
>   * not sure why we have them (backward compatiblity?) and how they ought to be used
>   * I added integration tests, feel free to let me know what from those tests are useless

They're meant for simple method to have inter-addon calls with a return value. I'm not sure if they're being currently used by any add-ons.

In the main addon (`script.example.callee`):
```python
def callback(data):
    returnCall('getData', 'test')

registerCallback('script.example.callee', 'getData', callback)
```

In the addon making the call to the main addon:
```python
x = makeCall('getData', source_id='script.example.callee')
```
x = 'test'

dagwieers commented:
@ruuk Hmm, that will be harder to test as both the sender and receiver are in the same namespace and as a result the receiver will also receive the signal it is returning to the sender. So you get an endless loop. I would need to redesign Addon Signals and the xbmc framework in order to make that work :-(

**Update**: So the endless loop was caused by the return signal containing a dot, and onNotification splitting by dots to get the signal name. So the return signal was identical to the received signal and the callback was continuously triggered.

dagwieers committed:
Fix a bug in the return signal

This PR includes:
- Fix a bug in the return signal name
- Improve callback integration tests
- Ensure receiver and sender use different SignalReceiver instances

dagwieers commented:
After fixing the bug in Addon Signal's return signal name, I think this is ready to be reviewed.

dagwieers commented:
Any news?

CastagnaIT commented:
I would have left import json at top level is the one mainly used
everything else is fine for me

dagwieers commented:
> I would have left import json at top level is the one mainly used

There is no need to import it when Addon Signals is imported given that imports are more expensive on slower platforms (Raspberry Pi). It is a good practice to not force it on others IMO.
https://github.com/pietje666/plugin.video.vrt.nu/wiki/Philosophy#importing-libraries-on-raspberry-pi

dagwieers commented:
@ruuk Any feedback on this?

dagwieers commented:
@ruuk Can we get this merged? Or at least provide a path forward with this?

dagwieers committed:
Move from Travis CI to GitHub CI

This commit includes:
- Add .gitattributes to get a proper ZIP pakage
- Remove Travis CI
- Add a GitHub CI workflow for unit testing, codecov and SonarCloud
- Add a GitHub CI workflow for kodi-addon-checker
- Move add-on icon to resources/

dagwieers commented:
I added GitHub CI support and removed Travis CI support. The unit tests now integrate better with GitHub and the output will look like this from the Actions tab at the top:
https://github.com/dagwieers/script.module.addon.signals/actions

It would be really nice if we can have the fixes and unit testing added.

ruuk commented:
> a curiosity question, i have seen that in all the projects you keep wrong the commas on the descriptions of the methods, why do you insert them all wrong?
> so you use the wrong ''' instead of """

Because I started writing Python before one style was decided to be the 'right' style, and only recently started caring :smile:.

ruuk reviewed (approved):
Sorry for the long, long delay in reviewing this. Changes look good, I'll merge after a rebase.

dagwieers commented:
Ok, I noticed a new release without any tests and then found this PR still in the queue :-(

Rebasing this is real hell, because I lack the context 10 months later and I have to dig into what has changed since. So it takes a lot more effort than it could have been. Please next time, if a big PR is in the queue, do not merge anything else before this has been sorted out.

dagwieers committed:
Fixes after rebase

dagwieers commented:
The current code is failing, where it used to work before:
```unittest
$ python -m unittest discover -v
test_bogus_sender (tests.test_calls.TestCalls) ... Iteration: 1
Iteration: 2
Iteration: 3
Iteration: 4
Iteration: 5
ERROR
test_calls (tests.test_calls.TestCalls) ... ERROR
test_send_multiple (tests.test_calls.TestCalls) ... ERROR
test_unicode (tests.test_calls.TestCalls) ... ok
test_addon_signals (tests.test_encoding.TestEncoding) ... ok
test_bogus_sender (tests.test_signals.TestSignals) ... ERROR
test_register_multiple (tests.test_signals.TestSignals) ... Debug: {u'test': u'register_multiple'}
ok
test_send_multiple (tests.test_signals.TestSignals) ... ok
test_send_none (tests.test_signals.TestSignals) ... ok
test_send_nonsignal (tests.test_signals.TestSignals) ... ok
test_send_unregistered (tests.test_signals.TestSignals) ... ok
test_signals (tests.test_signals.TestSignals) ... ok
test_unicode (tests.test_signals.TestSignals) ... ok
test_unregister_nonexisting (tests.test_signals.TestSignals) ... ok
test_warning (tests.test_signals.TestSignals) ... Notice: ++++==== script.module.addon.signals: sourceID keyword is DEPRECATED - use source_id ====++++
ok
```

dagwieers commented:
Ok, the new code returns `OSError` on abortRequested(). Not sure if this is the intended behaviour.
But after codifying this as the correct behaviour, the tests work as expected.

This is what it will look like in this project when merged: https://github.com/dagwieers/script.module.addon.signals/actions

dagwieers committed:
Add release workflow

The release workflow helps pushing releases to the Kodi repositories.
Also add support starting from Gotham

CastagnaIT commented on lib/AddonSignals.py:
```diff
@@ -22,102 +20,127 @@ def _perf_clock():
                     raise WaitTimeoutError
                 break
-            elif monitor.abortRequested():
+            if monitor.abortRequested():
```
try look here this was "elif"

dagwieers replied:
Sure, but pylint complains because it is actually not needed, the previous block always breaks.

dagwieers commented:
I just added a GitHub CI release workflow.

If you add `GH_USERNAME`, `GH_TOKEN` and `EMAIL`, you can push a new release to GitHub and Kodi repositories by simply tagging a new release and pushing it to GitHub.

CastagnaIT commented:
> Ok, the new code returns `OSError` on abortRequested(). Not sure if this is the intended behaviour.
> But after codifying this as the correct behaviour, the tests work as expected.
> 
> This is what it will look like in this project when merged: https://github.com/dagwieers/script.module.addon.signals/actions

now remember, yes the OSError raised is right behaviour
when Kodi stop the loop, the addon that have to receive the data
have to know when the operation is interrupted and stop his code

CastagnaIT commented:
if you prefeer you can add a new custom exception for interrupted event but is not really necessary

dagwieers commented:
> > Ok, the new code returns `OSError` on abortRequested(). Not sure if this is the intended behaviour.
> > But after codifying this as the correct behaviour, the tests work as expected.
> > This is what it will look like in this project when merged: https://github.com/dagwieers/script.module.addon.signals/actions
> 
> now remember, yes the OSError raised is right behaviour
> when Kodi stop the loop, the addon that have to receive the data
> have to know when the operation is interrupted and stop his code

Sure, but OSError seems to be a strange exception to be used for that. A custom exception, or maybe one from Kodi would make more sense IMO. A few months ago this was not handled, and did not raise an OSError.

CastagnaIT commented:
if you want i think you can still add a custom exception
this code is not yet released

dagwieers commented:
@CastagnaIT Sure, but I would leave that up to the person who wrote that ;-)
I would like to get this one merged first.

dagwieers commented:
The coverage is no longer 99%, this is something to fix later IMO
https://codecov.io/gh/dagwieers/script.module.addon.signals/src/assorted-fixes/lib/AddonSignals.py
